hammerblade/torch/kernel/bsg_tensor.hpp

```diff
@@ -122,6 +122,11 @@ class BSGTensor {
 
       return data[offset];
     }
+
+    DT& operator[](uint32_t n) {
+      bsg_assert_msg(n < N, "error: N=%d but accessed %d\n", N, n);
+      return data[n];
+    }
```

Comment:
I guess what you are trying to do here is similar to this case:

https://github.com/cornell-brg/hb-pytorch/blob/8dc12f36bdbc84c4a6b46504b97556f8d1ba7699/hammerblade/torch/kernel/bsg_tensor.hpp#L102-L109

I think @cbatten prefers to use `[]` for tensor accesses. So how about we make the special case , which we assume data is contiguous and index into it with linear indices, `()`, and regular ones `[]`?

Reply:
Thought `[]` should be plain C++ array indexing and `()` could be for special indexing? What do you think?

If you still think we should swap definitions, can I do it in a separate pr as it needs modifications to many kernel definitions as well?

Reply:
FWIW, I would tend to agree with @cbatten's implied suggestion here: use the "right" `operator` override for the job. Parentheses generally imply "calling it like a function" to most programmers, whereas square brackets generally imply "subscripting" to most programmers.

The general aesthetics of operator overloading is that you should be able to use the same "verb" for different "nouns": that is, `foo[bar]` should roughly "mean" the same thing regardless of the type of `foo`. Its implementation is decided by overloading, but the programmer should not have to remember to do different things depending on the type. That's the compiler's job!

Reply:
I think it makes sense... 

Even better, I think, we can just have `[]` for indexing. Something like:
```
template<typename... T>
DT& operator()(T... indices) {
  std::initializer_list<uint32_t> iarray = {indices...};

  if (iarray.size() == 1) {
    // implementation for linear indexing: a[5]
  } else {
    // implementation for MDA indexing: a[1, 5]
  }
}
```

Reply:
> I guess what you are trying to do here is similar to this case:
> 
> https://github.com/cornell-brg/hb-pytorch/blob/8dc12f36bdbc84c4a6b46504b97556f8d1ba7699/hammerblade/torch/kernel/bsg_tensor.hpp#L102-L109

Oh, linear indexing is already implemented! Sorry, for some reason, I didn't realize this before

Reply:
> I think @cbatten prefers to use `[]` for tensor accesses. So how about we make the special case , which we assume data is contiguous and index into it with linear indices, `()`, and regular ones `[]`?

@cbatten I recall my preference was `[]`, but it's hard to do that in C++.

Reply:
() is not a good operator for indexing. Why is using [] hard to support?

Reply:
I could imagine we could support multidimensional indexing with [] and flattened linear indexing with a method call for example at() ... maybe

Reply:
Adrian is a great C++ hacker. Maybe he has some ideas on how we can use template meta programming to simplify both flattened element wise kernels (like relu) as well as kernels like addmm and sum